### ISSUE TITLE
[@typescript-eslint/restrict-plus-operands] False positive for type `0|1`

### ISSUE THREAD

**Issue by viceice** [typescript-eslint/typescript-eslint#4689]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": ["error"]
  }
}
```

```TS
export function getFutureVersion(
  policy: number,
  newVersion: string,
  baseVersion?: string
): number[] {
  const toRelease: number[] = [];
  const baseRelease: number[] = [];
  return baseRelease.map((_, index) => {
    const toPart = toRelease[index] || 0;
    if (index < policy) {
      return toPart;
    }
    if (index === policy) {
      return toPart + (baseVersion === undefined ? 0 : 1);
    }
    return 0;
  });
}
```

- Reproduction [Playground](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLCHRLH7k4wCuUZA9noQIYIAmUVAC4wAbVQRChDoVQBdMCGABYAFDqIADxyyUAMxYJGcDgjABzCEgBiLJC1oA1KVFMIAFOrBgARjygIF0I3MwB+USh6RAt1AEpRBBYAW18pMUUAb28wfDMosCQOACUIGAgAiESUtMIMsABecXkAbhy8hAL-QNLyyurU9MUmjLa1H1oHQnNuiF6KwIA6ZJ4cDw8AfQAaMER+bTjGgD4wbPGfXPyUIoAFHkIUJqL5yrE97UUwsLAABjGL3b6MAed5aMAAHjAAEZDmcARNrI5zLd7kh-hcAL45HxwIEggTaRoNJow07Yi6TJGFDh3B5gADUwNmwVC5mJTSM+30iAg-DA3x+YFEMPRPix5wRU3MfxyGLiY3FQA&rules=N4XyA&tsConfig=N4XyA)
- Rproduction [renovate-reproductions/typescript-eslint-issue-4689](https://github.com/renovate-reproductions/typescript-eslint-issue-4689/blob/0f5918e579a1126fa27e105d6ffab5955a423c2a/index.ts#L10)

happens after update from `v5.12.1`to `v5.13.0`.
- renovatebot/renovate#14563
- renovatebot/renovate#14597

**Expected Result**

Should not cause an error. 

**Actual Result**
![image](https://user-images.githubusercontent.com/1798109/158336099-54be5643-1c16-488f-abb7-3ccc08f96a97.png)
The rule causes a false positive on code above.
https://github.com/renovatebot/renovate/pull/14597/files#diff-f6e993a4715dd7526818c5bb2cdfdaa3a139db2aa290ba8857b9d93f8fe03c4f

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.14.0` |
| `@typescript-eslint/parser`        | `5.14.0` |
| `TypeScript`                       | `4.6.2` |
| `ESLint`                           | `8.10.0` |
| `node`                             | `14.19.0` |

**Comment by bradzacher** [comment]
I am unable to repro this against master.
[playground](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLCHRLH7k4wCuUZA9noQIYIAmUVAC4wAbVQRChDoVQBdMCGAAoFfg4IoKAG6iADAB8AjGAC8YYwG51m7WB2mLpgNQObQA&rules=N4XyA&tsConfig=N4XyA)

**Comment by vpzomtrrfrt** [comment]
I'm also seeing a false positive with this rule that I can't reproduce in a new project. Seems to be related to the upgrade from typescript 4.5 to 4.6

**Comment by bradzacher** [comment]
This rule hasn't changed in several months, definitely not so in the latest version. So it's doubtful that it was the latest release that broke something.

If you can create an isolated repro in either the playground or in a stand-alone git repo, we can definitely help out and take a closer look!

Without a repro - we unfortunately can't provide much help.

**Comment by viceice** [comment]
> I am unable to repro this against master. [playground](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLCHRLH7k4wCuUZA9noQIYIAmUVAC4wAbVQRChDoVQBdMCGAAoFfg4IoKAG6iADAB8AjGAC8YYwG51m7WB2mLpgNQObtrSiR0AYjzgYc0swNwByMJsgA&rules=N4XyA&tsConfig=N4XyA)

I see the error on the playground
![image](https://user-images.githubusercontent.com/1798109/158559079-9f801032-0d5c-477e-8fd1-aabff485d7df.png)

**Comment by bradzacher** [comment]
You see the test error I added to show the rule is working.. You don't see an error on the code that from your repro.

**Comment by viceice** [comment]
Updated repro code. Not sure how to reduce more. If you copy it to the playground, it will show the issue.

**Comment by viceice** [comment]
it seems `toPart` isn't correctly inferred
![image](https://user-images.githubusercontent.com/1798109/158635037-64b9de1e-6b39-4dec-8fea-ce28063cea00.png)

**Comment by viceice** [comment]
OK, verified. This works when explicit typing `toPart`
![image](https://user-images.githubusercontent.com/1798109/158637405-48984e68-3579-4183-b197-3ad672259494.png)

**Comment by viceice** [comment]
Reproduction playground added to issue

**Comment by JoshuaKGoldberg** [comment]
Thanks for following up with more details @viceice! The issue in the playground code snippet is that `toPart` is type `any`. If we change the playground code snippet to `const toPart: number` there's no more issue. [updated playground link](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLCHRLH7k4wCuUZA9noQIYIAmUVAC4wAbVQRChDoVQBdMCGABYAFDqIADxyyUAMxYJGcDgjABzCEgBiLJC1oA1KVFMIAFOrBgARjygIF0I3MwB+USh6RAt1AEpRBBYAW18pMUUAb28wfDMosCQOACUIGAgAiESUtMIMsABecXkAbhy8hAL-QNLyyurU9MUmjLa1H1oHQnNuiF6KwIA6ZJ4cDw8AfQAaMER+bTjGgD4wbPGfXPyUIoAFHkIkAdrGwpKyhYgxPe1FMLCwAAMYwuu30YA83y0YAAPGAAIyHM4gibWRzmW73JDAi4AXxyPjgYIhAm0jQaTQRp3xF0maNedweYAA1ODZsFQuZyU0jPt9IgIPwwP8AWBRAjsT48ecUVNzECcji4mMpUA&rules=N4XyA&tsConfig=N4XyA)

By the way, as a general tip, instead of posting screenshots of code or CI reports, please post plain text and/or links to playground snippets. Screenshots are harder for us to work with: they can't be copy&pasted and aren't accessible to folks who rely on screenreaders unless they have very detailed alt text / captions.

**Comment by viceice** [comment]
Sorry about the pictures, just want to show the issue. 

I know that adding `const toPart: number` will fix it, but `toPart` should already be number. 
vscode show's it as `number`:
![image](https://user-images.githubusercontent.com/1798109/158643689-e230a47e-526a-4b7b-9ca7-823e87afcbb0.png)

Will add a git repo for reproduction too. https://github.com/renovate-reproductions/typescript-eslint-issue-4689/blob/0f5918e579a1126fa27e105d6ffab5955a423c2a/index.ts#L10

**Comment by bradzacher** [comment]
This is a bug with TS.

You can repro it [in the TS playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&useUnknownInCatchVariables=true&noFallthroughCasesInSwitch=true&noImplicitOverride=true#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwObAwBiSMSUwAasFAM7YoAUARgIa1U304D8AXHFowomFHgCUAlEgC2zGgG0AunADeAKDhw0OIXBgQASsAA2wdsCmz5UZXAC8cZQG5N23fDYdjZi1bmKKo4ubhRkULhewD7mHAB0MqxgjIwA+gA0cKIAJqDiDgB8am5aOih6BgAKrLAO+kamscAKOaAqPDxwAAyuWlqYCHCMrSBwADxwAIz5Gn19YeS4VTUwvX0AviVZg8MouaP2h1MzW1oLEfXVtQDUQ1HUdAwOR6i5CKLA2XCdXXAC0844FtNnNzrgem51uJXOsgA) - TS intellisense shows that `toPart` is `any` in v4.6.2.

But at the same time TS knows that the type of `toPart` is a number:
![image](https://user-images.githubusercontent.com/7462525/158650009-26d8e3ce-1bf8-477a-8ec6-85ad3bed6627.png)

However in TS v4.5.5 this does not repro - it shows that the type is `number`.

In TS 4.6.2 the API that we use to get the type (`checker.getTypeAtLocation`) returns the type as `any`, which is all we can base our linting off of.

This is a bug in TS.

**Comment by viceice** [comment]
We had ts 4.6.2 with both ts-eslint versions v5.12.1 and v5.13.0 and it started failing on v5.13.0. I see v5.14.0 has added ts v4.6 support.

So it's strange to me why it fails on v5.13.0 🤔

**Comment by viceice** [comment]
@bradzacher Can you help me to find or file a typescript bug? I'm not really sure how to do this, as the normal typescript compiler doesn't show any issue.

**Comment by viceice** [comment]
Found another sample of failing type inference
```ts
/* eslint "@typescript-eslint/restrict-template-expressions": ["error"] */
const PREFIX_DOT = 'PREFIX_DOT';

const TYPE_NUMBER = 'TYPE_NUMBER';
const TYPE_QUALIFIER = 'TYPE_QUALIFIER';

export interface BaseToken {
  prefix: string;
  type: typeof TYPE_NUMBER | typeof TYPE_QUALIFIER;
  val: number | string;
  isTransition?: boolean;
}

export interface NumberToken extends BaseToken {
  type: typeof TYPE_NUMBER;
  val: number;
}

export interface QualifierToken extends BaseToken {
  type: typeof TYPE_QUALIFIER;
  val: string;
}

export type Token = NumberToken | QualifierToken;

export function tokensToStr(tokens: Token[]): string {
  return tokens.reduce((result, token, idx) => {
    const prefix = token.prefix === PREFIX_DOT ? '.' : '-';
    return `${result}${idx !== 0 && token.val !== '' ? prefix : ''}${
      token.val
    }`;
  }, '');
}
```

repro added to https://github.com/renovate-reproductions/typescript-eslint-issue-4689

[Playground](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLCHRLH7lIQC2OMAhsxQB461QocAPYIoqAFxgA2qgiFCwwqgC6YEMACwAKB35RUFAAUASgFEAYgEkAGgH0AIgHkAKmAC8YAOSnLtx65eANw6egYoLgCaRmZ2AHIAqgCyAEJmJh7eUTHxyWkmwWFiEdGxAIoJAIIAMlbW6Zle2eVVtfUFIbraEHxKKIjMhABmHPgQYCkcUBAuwgDWEAhgAN46YGD8EENwPFKGxAgA5p3r2HhSZxDCQ2DNuakNAD5gl9e3pXYVNXVW6SdgADcODApAgAK4sABG8jAz32iGOazAcCgLkIHDEcCQIgQAH4pJDhMIYBAMZ0AL6hbq9Qj9ZDyEZjMBxCHQwizBZLHrMBAAEygEymM3mixWSMuF1wVxud0SDxM-yBILA4Kh8gpVJ6OD6yPpw1G4zKYOBcG28g5ou5i35gumFqWq20pylkrwbzuXzavwVSKVe3oCI1XS1OsutxFS08LLV7IjsLARpNZtjnM6OiGYIQjBxLwjqOEAGV6AAKJB5qT26QqACU-oOhzFTrAtCQYMISzLnKgADpaLywWNi8WBGCYEgADS5zmTuC8njVjwAPkb63W+mKG1o2x4mU7i27m23Hncnl81nszjcuO83a8YCkXjIhSb6xbbaWAAMACTLEdj8k-rOO4AIQnmAAAMYAAGRQVO+5KmAoGeF4d7XoeOz3t4XgAY6q6rnuCDdkqSLrOSH7-OSk4odWFJAA&rules=N4XyA&tsConfig=N4XyA)

**Comment by bradzacher** [comment]
So I simplified your repro further to this eslintrc
https://github.com/bradzacher/typescript-eslint-issue-4689/blob/patch-1/.eslintrc.js

The interesting thing is that if you remove the lint rule `'@typescript-eslint/no-misused-promises': ['error'],` from the config, the error in `restrict-template-expressions` goes away!!!

**Comment by viceice** [comment]
I've reduced the repo at https://github.com/renovate-reproductions/typescript-eslint-issue-4689 to be very minimal.

@JamieMagee is working to reduce it even futher and will open an upstream issue on typescript.

it seems this additional lint rules make a difference config
```js
  rules: {
    // TODO: removing these two lines to remove the lint errors
    '@typescript-eslint/no-unsafe-return': 0,
    '@typescript-eslint/no-unsafe-argument': 0,

    '@typescript-eslint/restrict-template-expressions': [
      2,
      { allowNumber: true, allowBoolean: true },
    ],
    '@typescript-eslint/restrict-plus-operands': 2,
  },
```

https://github.com/renovate-reproductions/typescript-eslint-issue-4689/blob/3432b05c5700c04a132f41e092677da566bd7796/.eslintrc.js#L17-L28

**Comment by viceice** [comment]
> So I simplified your repro further to this eslintrc https://github.com/bradzacher/typescript-eslint-issue-4689/blob/patch-1/.eslintrc.js
> 
> The interesting thing is that if you remove the lint rule `'@typescript-eslint/no-misused-promises': ['error'],` from the config, the error in `restrict-template-expressions` goes away!!!

So it seems there is some crossing between those rules, which cause the strange behavior. We do not even have any `Promise` reference in example code 😕

**Comment by bradzacher** [comment]
None of our lint rules mutate any of the TS data - they all just read TS data.
I confirmed this by adding `readonly` to all of the TS types and I got no errors

All TS types are computed lazily - TS computes exactly what you ask for and nothing more! So if I had to hazard a guess, what's happening is that there's a bug in TS wherein it is *not* computing the correct type information for the variable based solely on the query in `restrict-template-expressions`. `no-misused-promises` must query for some additional types which then changes the types that TS returns when `restrict-template-expressions` runs after it.

I went through and commented out all of the code in `no-misused-promises` until it stopped erroring. I narrowed it down to the [`VariableDeclarator`](https://github.com/typescript-eslint/typescript-eslint/blob/cf325cea6066181563e2889890eb1c54b294bdd9/packages/eslint-plugin/src/rules/no-misused-promises.ts#L151) selector - if I comment this one out the errors disappear.

Next I commented out code from within its function. When I [comment out this line](https://github.com/typescript-eslint/typescript-eslint/blob/cf325cea6066181563e2889890eb1c54b294bdd9/packages/eslint-plugin/src/rules/no-misused-promises.ts#L244) the errors disappear.

https://user-images.githubusercontent.com/7462525/158867241-934b12d7-848b-4443-b38a-ea96936bf3e8.mov

**Comment by bradzacher** [comment]
I have raised https://github.com/microsoft/TypeScript/issues/48313 which covers one of the issues and was easy to isolate just using the playground.

**Comment by Josh-Cena** [comment]
On the latest TS version, this doesn't seem to be an error anymore.
